packages/csvviewer/src/model.ts

```diff
@@ -0,0 +1,663 @@
+} from './parse';
+
+/*
+Possible ideas for further implementation:
+
+- Show a spinner or something visible when we are doing delayed parsing.
+- The cache right now handles scrolling down great - it gets the next several hundred rows. However, scrolling up causes lots of cache misses - each new row causes a flush of the cache. When invalidating an entire cache, we should put the requested row in middle of the cache (adjusting for rows at the beginning or end). When populating a cache, we should retrieve rows both above and below the requested row.
+- When we have a header, and we are guessing the parser to use, try checking just the part of the file *after* the header row for quotes. I think often a first header row is quoted, but the rest of the file is not and can be parsed much faster.
+- autdetect the delimiter (look for comma, tab, semicolon in first line. If more than one found, parse first row with comma, tab, semicolon delimiters. One with most fields wins).
+- Toolbar buttons to control the row delimiter, the parsing engine (quoted/not quoted), the quote character, etc.
+- Investigate incremental loading strategies in the parseAsync function. In initial investigations, setting the chunk size to 100k in parseAsync seems cause instability with large files in Chrome (such as 8-million row files). Perhaps this is because we are recycling the row offset and column offset arrays quickly? It doesn't seem that there is a memory leak. On this theory, perhaps we just need to keep the offsets list an actual list, and pass it into the parsing function to extend without copying, and finalize it into an array buffer only when we are done parsing. Or perhaps we double the size of the array buffer each time, which may be wasteful, but at the end we trim it down if it's too wasteful (perhaps we have our own object that is backed by an array buffer, but has a push method that will automatically double the array buffer size as needed, and a trim function to finalize the array to exactly the size needed)? Or perhaps we don't use array buffers at all - compare the memory cost and speed of keeping the offsets as lists instead of memory buffers.
+- Investigate a time-based incremental parsing strategy, rather than a row-based one. The parser could take a maximum time to parse (say 300ms), and will parse up to that duration, in which case the parser probably also needs a way to notify when it has reached the end of a file.
+- For very large files, where we are only storing a small cache, scrolling is very laggy in Safari. It would be good to profile it.
+*/
+
+/**
+ * Possible delimiter-separated data parsers.
+ */
+const PARSERS: {[key: string]: IParser} = {
+  'quotes': parseDSV,
+  'noquotes': parseDSVNoQuotes
+};
+
+/**
+ * A data model implementation for in-memory delimiter-separated data.
+ *
+ * #### Notes
+ * This model handles data with up to 2**32 characters.
+ */
+export
+class DSVModel extends DataModel implements IDisposable {
```

Comment:
Several methods/getters are lacking a return type. Is that on purpose?

Reply:
No, thanks for catching that.